### DISCUSSION TITLE
[Lua] contextmenu_gui 图形化右键菜单

### DISCUSSION THREAD

**Discussion by hooke007** [hooke007/mpv_PlayKit#60] [INFO 信息]
# 本文已过时

参见 https://github.com/hooke007/MPV_lazy/commit/5c26025e2e92f2a7625900e77e21ee45a18a2c57

- 如果需要独立于mpv窗口外的图形菜单
推荐改用原生的上下文菜单功能 https://mpv.io/manual/master/#context-menu
原始的菜单构建方法需要自行编写Lua脚本
  - 也可以使用辅助脚本 contextmenu_plus （或上游脚本 dyn_menu ） 进行基于 input.conf 文件的简易快速构建

- 如果只需要融合进画面的菜单，建议使用脚本 uosc 附带的OSD菜单 https://github.com/hooke007/MPV_lazy/discussions/186
它也可以基于 input.conf 文件进行简易快速构建

<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>

# 简介

<img src="https://user-images.githubusercontent.com/41094733/145067213-b80c95dd-0ee2-4374-a587-48b8eebe875c.jpg" width="400"> <img src="https://user-images.githubusercontent.com/41094733/145070058-71203e85-5766-4cf2-886a-b0ff1bf9823e.jpg" width="400">

contextmenu_gui 是一组实现mpv的图形化右键菜单的脚本群。

>[!CAUTION]
>**菜单功能也可由 [uosc](https://github.com/hooke007/MPV_lazy/discussions/186) 提供。推荐改用 uosc**

该脚本组修改自 https://gitlab.com/carmanaught/mpvcontextmenu  
修改适配了（且仅支持）win10/win11，经过摸索测试后最终实现了本地化和便携化。（Linux用户如有该需求请使用上游的原版）
<s>预设的键位结构大体依照的是 [mpv.net_CM](https://github.com/hooke007/mpv.net_CM#预览) 的右键菜单</s>

_只有 **mpv release 0.33.0** 之后的版本才具有支持脚本群组的特性_  
<s>_使用 **懒人包2021EX** 及之后版本的用户直接跳过中间章节，从自定义配置起开始阅读_</s>
_mpv-lazy已于 **obs** 版本之后中止集成此脚本_

# 必要前置

1. tcl工具包（必须含 `tk` 模块） http://www.tcl-lang.org/software/tcltk/bindist.html
该组件负责运行时绘制菜单的图形框架。
不是所有预编译的版本都能正常使用，你可能会遇到菜单文本乱码，菜单启动报错等问题。（经过我长时间测试过的可用的是：v8.6.x）
**ActiveTcl**（完整sdk，体积较大，零散文件很多） https://www.activestate.com/products/tcl/
**KitCreator**（精简库，体积很小，单文件） http://tclkits.rkeene.org/fossil/wiki/Downloads
单文件版的可以直接解压到 `mpv.exe` 所在目录，安装版的自行提取或直接安装。

>[!TIP]
>如果不清楚下哪个，那就从 👉 [**此处**](https://github.com/hooke007/dotfiles/releases/tag/mpv_addones) 获取现成的单文件版。

2. 下载脚本群组的文件夹到mpv的设置文件夹（看不懂文字说明就看我的repo目录结构）
即 `contextmenu_gui` 文件夹位于 `scripts` 内。对应的脚本设置文件与一般脚本的放置位置无差别

# 可选的关联脚本

<img src="https://user-images.githubusercontent.com/41094733/145069466-4e60cb2c-8359-4cb0-b5af-a77fdf1c99da.jpg" width="400">

一些预置键位的功能由其它第三方脚本提供，在菜单内已标注 `【外置脚本】`

1. **load_plus.lua**
2. **playlist_osd.lua**

https://github.com/hooke007/MPV_lazy/tree/main/portable_config/scripts  
包括对应设置文件，下载到正确位置

# 使用说明

编辑你的 **input.conf**  
内置此参数 绑定一个按键后可在mpv内唤起菜单
```
MOUSE_BTN2 script-message-to contextmenu_gui contextmenu_tk
```
_不会有人不喜欢鼠标右键吧_  
最好不要同时使用 `MOUSE_BTN2` 和 `MOUSE_RIGHT` （正常鼠标这两都是指向都是右键）

## 菜单选项

[**contextmenu_gui_engine.conf**](https://github.com/hooke007/MPV_lazy/blob/main/portable_config/script-opts/contextmenu_gui_engine.conf) 该文件修改菜单所用的默认字体
其中，必须保证 `tcltkBin` 指向你实际下载使用的tcl/tk二进制文件 默认值为 `tclsh` （无需扩展名，可用绝对路径）

如果你使用单文件版（例kitcreator的的当前版本）放在 **mpv.exe** 所在目录， `tcltkBin` 的值 只需匹配文件名 `tclkit-gui-8_6_12-twapi-4_6_0-x64-max`， 或者改二进制的文件名为 `tclsh` 而无需修改脚本设置文件。

## 自定义配置

[**contextmenu_gui.conf**](https://github.com/hooke007/MPV_lazy/blob/main/portable_config/script-opts/contextmenu_gui.conf) 提供了可自定义的空档位滤镜/着色器各十个  
根据我的示例和说明进行适配你的本机修改。

<s>_对应快捷键的参数只是纯粹的文本显示_</s>  （此功能已砍）
可能存在语法上的问题，<s>比如无法并列多个着色器，</s>图上红圈部分是仅供示范文本显示，具体阅读示范配置文件的说明。  
<img src="https://user-images.githubusercontent.com/41094733/145068779-b2705312-ab11-4d34-b91f-3de33e131b32.jpg" width="600">

# 其它

- 未载入文件时，菜单只显示很少的选项
<img src="https://user-images.githubusercontent.com/41094733/145067993-b2818fbf-f6ee-4c6d-b017-02be608678ab.jpg" width="500">

- 冷启动菜单的速度不快，但后续唤起延迟应该很低

- 不同tcl/tk编译版的性能和其它反馈表现可能不一致

- 一些键位（例如均衡器）会二次唤起菜单，原本的设计是直接弹出到上次点击的位置，但是不知道为什么在windows上只能显示1级菜单

- 应该几乎没人在windows上广泛使用过这套脚本，可能存在各种潜在问题，欢迎当白鼠（我大概没能力解决，上游只在linux上使用，没有技术支援）

**Comment by Youxikong** [comment]
反映个问题，设置d3d11全屏独占后（其他设置与原版一致），开启全屏后，打开右键菜单会导致视频画面暂停，右键彩蛋关闭、退出全屏，依然是静止的，必须前进后退画面才会运动起来。关闭全屏独占的时候全屏下是正常的。

**Reply by hooke007** [reply]
反正没事别开独占没啥优势。而且看起来不像是菜单的问题是独占/驱动的问题，唤起菜单后焦点变成了菜单而不是mpv

以及这个选项有没有尝试
https://github.com/hooke007/MPV_lazy/blob/e4c5fd9e7c0ec7aa57b9a488d1dab7fdebd6dcb7/portable_config/mpv.conf#L20

**Reply by Youxikong** [reply]
尝试了，并没有效果，还是冻结了。

**Reply by hooke007** [reply]
我这里没有问题

**Reply by Youxikong** [reply]
登登这个应该写no吗？我写了yes。😅等下再看看。

**Reply by hooke007** [reply]
默认就是yes，写yes跟没改没区别

**Comment by dyphire** [comment]


**Reply by hooke007** [reply]
全角字符是故意的，和半角区分种类。。至于等宽。。我也不是太在意这个（而且conf里用户也能自己改
tcl不读fonts也是正常的，毕竟不是一个程序，fonts那个功能是给字幕功能用的

**Reply by hooke007** [reply]
见仁见智吧。不会改默认值，用户级设置已经够方便了

**Reply by hooke007** [reply]
没必要啊。。这个效果本身就是等宽的特性不是tcl这个程序的特性

**Reply by hooke007** [reply]
理由已说明，不再复述。（补充：字体是主观性极强的东西，谁的建议都是主观性建议，我不喜欢用对不对来定义非等宽的显示；文本对齐是等宽字体的常识，为什么要注释多余的废话....至于键位右对齐，无感

此话题终结。

**Comment by sunjanesy** [comment]
我试了一个钟头，请教下我是哪里设置不对，无法调用新版菜单
![新菜单3](https://github.com/hooke007/MPV_lazy/assets/104602119/123c35f8-0c2b-46f4-9021-9910d16fb8c9)

**Reply by hooke007** [reply]
**必须包含tk模块**
另外新版菜单又是什么？

**Reply by sunjanesy** [reply]
> **必须包含tk模块** 

![新菜单3](https://github.com/hooke007/MPV_lazy/assets/104602119/123c35f8-0c2b-46f4-9021-9910d16fb8c9)
这个不是大佬推荐的tk单文件么。8.6.x

>另外新版菜单又是什么？

就指的这篇文章的菜单（相对于原版uosc菜单）

其实是因为只有这个才能配合[osc_plus.lua](https://github.com/hooke007/MPV_lazy/blob/main/portable_config/scripts/osc_plus.lua)，不然我用原版uosc一整套，估计又要被批，明明是完全冲突的两套。

**Reply by hooke007** [reply]
你看发帖时间也应该知道uosc的是新菜单
你这是不带tk的tcl

我也没看懂 只有这个能配合input_plus 是什么意思，我input_plus就是uosc之后写的，uosc之后我再也没实际用过这个旧菜单

**Reply by sunjanesy** [reply]
> 你看发帖时间也应该知道uosc的是新菜单 你这是不带tk的tcl
 
文本太多了.真没注意先后时间
> 我也没看懂 只有这个能配合input_plus 是什么意思，我input_plus就是uosc之后写的，uosc之后我再也没实际用过这个旧菜单

刚刚手滑粘贴错了[osc_plus.lua](https://github.com/hooke007/MPV_lazy/blob/main/portable_config/scripts/osc_plus.lua)
因为没法单独调用uosc的菜单，才想着和项目首页截图一样搭配

**Reply by hooke007** [reply]
“obs 版本之后中止集成此脚本” 只看正文也写了obs版本之前的啊

**Reply by sunjanesy** [reply]
> “obs 版本之后中止集成此脚本” 只看正文也写了obs版本之前的啊

咱真没办法每句话都去细品参透，我的初衷就是给[osc_plus.lua](https://github.com/hooke007/MPV_lazy/blob/main/portable_config/scripts/osc_plus.lua)加一个不冲突的右键菜单。试试看

**Reply by hooke007** [reply]
我写文档和后续的修正都尽量不写废话，你的主要问题的原因就是没看仔细。。。

**Reply by sunjanesy** [reply]
> 我写文档和后续的修正都尽量不写废话，你的主要问题的原因就是没看仔细。。。

有些东西真是不懂，就好比“必须带tk”我明显看了又看的，但是下载文件时，感觉都差不多，经提醒才知道说的是下面有一栏带Tk标识的。
总算在大佬提醒下召唤出来白色无乱码的菜单。离计划又近了一步

**Reply by hooke007** [reply]
不要在文档下报告问题

**Comment by tsl0922** [comment]
https://github.com/mpv-player/mpv/issues/5500#issuecomment-1849429931

这是一个基于 [C PLUGINS](https://mpv.io/manual/master/#c-plugins) 实现的 windows 右键菜单插件，好处就是使用原生 win32 api 实现，速度和体验和原生 windows 程序一样。

只需要把编译后的 dll 文件放到 `scripts` 目录即可被 mpv 自动加载，效果：

<img width="721" alt="image" src="https://github.com/mpv-player/mpv/assets/1680515/66aa6093-6b2b-4e34-a72d-c32466fbf435">

~~（实验性代码，仅验证想法用，目前写死了菜单项，还不支持从文件加载）~~

**Reply by dyphire** [reply]
基于 C 插件实现右键菜单的想法很不错，感觉只需要优化两点就很好了：黑暗模式或更好的菜单渲染（类似 mpv.net），可以自定义菜单项

**Reply by tsl0922** [reply]
黑暗模式不需要再实现，因为使用的原生 win32 api，跟随系统自动支持了。

**Reply by dyphire** [reply]
那除了自定义菜单项(最好实现从 input.conf 读取)外，还有两个可以增强的地方：实现像 uosc 菜单一样按住 shift 点击菜单操作后不关闭菜单，对于一些重复性操作比较好；实现 mpv.net 的轨道菜单的功能，播放器菜单的标配了

**Reply by tsl0922** [reply]
你说的都可以实现，就是需要时间折腾。甚至理论上来说 mpv.net 的所有功能几乎都可以用 c 插件的方式实现，mpv.net 作者应该也有计划制作类似这种 C 插件版本的 mpv.net。

通过这种方式实现 mpv 图形界面应该算是除 mpv 自己实现外最为完美的了，因为不管是 mpv.net 这种 `wid` 嵌入式还是 `libmpv` 都有一个共同的问题：需要自己实现窗口，要想兼容 mpv 那么多的窗口选项还是需要做很多工作的，而且 `libmpv` 方式在渲染上有先天残疾，只能用 OpenGL（[ImPlay](https://github.com/tsl0922/ImPlay) 已经沦为玩具，快不想继续折腾了 😞 ）。

**Reply by dyphire** [reply]
是的，我现在不在用 mpv.net 的原因就是 mpv 上游很多新增的窗口特性它都还没添加支持，而且用 net.6 重写后启动时窗口会有大小变化的过程，应该是 bug

**Reply by tsl0922** [reply]
加了 actions 脚本，可到 https://github.com/tsl0922/mpv-plugin/actions 下载编译好的 `menu.dll` 体验了。

**Reply by zhongfly** [reply]
C PLUGINS的window支持还是mpv.net作者提的请求，https://github.com/mpv-player/mpv/issues/12185 到底有哪些功能是必须要通过前端来实现呢？

**Reply by tsl0922** [reply]
已经支持从 `input.conf` 读取 mpv.net 格式的菜单配置。

**Reply by dyphire** [reply]
目前是直接绑定在了右键上？我测试了一下，发现与 uosc 的菜单语法不兼容（倒不是什么问题），但右键会唤醒两个菜单就有点问题了

顺便我认为黑暗模式还是需要可控制，mpv.net、mpc-hc 和 mpc-be 的 UI 菜单都是可以设置始终黑暗模式的（我更喜欢这种）。但我觉得应该在 mpv 上游实现相关选项，在插件中读取选项状态来跟随 mpv 的黑暗模式行为，选项不存在时回退系统状态

**Reply by dyphire** [reply]
我看到你刚刚添加了 uosc 的`#!`绑定的菜单语法，不过 uosc 其实也支持 mpv.net 的`#menu:`菜单语法。我指的不是这个问题，因为我测试用的就是后一种兼容的菜单语法，在我的测试中 uosc 的同一个一级菜单项的子菜单内容在 menu.dll 的菜单中不会合并为一个主菜单，不清楚出了什么问题，我的 [input.conf](https://github.com/dyphire/mpv-config/blob/master/input.conf) 供测试排查

**Reply by tsl0922** [reply]
我测试一下，可能是文件编码问题，开发时用的 mpv.net 那个配置，里面没中文。

**Reply by tsl0922** [reply]
已修复，改用 unicode 版本的 win32 api 接口即可（界面显示的文字都做了编码转换）。

**Reply by zhongfly** [reply]
有点离题：有没有可能之后把ImPlay里的那个调试/信息工具独立出来？

**Reply by Youxikong** [reply]
读取input.conf貌似是写死了`~~/input.conf`，修改配置`input-conf`也不会从这里读取。

**Reply by tsl0922** [reply]
不是很想支持自定义 `inpu-conf`，因为 mpv 的路径逻辑太复杂了，就连现在写死的这个逻辑也是偷懒做法。

**Reply by tsl0922** [reply]
ImPlay 里的调试工具原本就有计划独立弄出来，有时间再折腾下吧。

**Reply by dyphire** [reply]
> 不是很想支持自定义 `inpu-conf`，因为 mpv 的路径逻辑太复杂了，就连现在写死的这个逻辑也是偷懒做法。

支持`input-conf`可能还是有些意义的，比如目前支持 mpv.net 和 uosc 的菜单语法，但两者的分割线写法并不相同。一些用户可能希望为 menu 和 uosc 的菜单指定不同的 input.conf 文件来使用各自的菜单功能

当然最简单的方法应该是添加对 uosc 的菜单分割线语法的支持

**Reply by tsl0922** [reply]
~~uosc 菜单分割线语法是支持的，uosc 是 `---`，实际上检测到第一个字符是 `-` 就会判定为分割线了。~~(不对，好像还是需要单个`-`）

不支持的是 uosc 那个省略 key 的写法，这样整行都是注释了，目前会被忽略。

**Reply by tsl0922** [reply]
关于 `input-conf` 支持，可以做。已经找到了一个简单点的实现方式，可以使用 mpv 的  `expand-path` 命令来解析路径。

**Reply by dyphire** [reply]
~~uosc 那个省略 key 的写法实用性很低，实际写菜单时基本用不到，我感觉没必要支持~~

**Reply by hooke007** [reply]
我不这么认为，实际上我已经不用快捷键很久了

**Reply by Youxikong** [reply]
我是能用菜单就用菜单。某种程度上，控制台输命令都比死记硬背快捷键方便一些。😶

**Reply by dyphire** [reply]
不一样，这里的描述不对，uosc 那个语法不是省略 key，是省略了 key 和 command，就剩注释的菜单语法。~~省略 key 的语法是常规菜单语法了，mpv.net 一样支持~~ 太久没用 mpv.net 了，都忘记它的菜单语法了

**Reply by dyphire** [reply]
不说我都没有注意到这个插件不支持省略 key 的语法，我刚才以为是后一种纯注释语法不支持。这感觉实用性就不太够了，我也很喜欢写省略 key 的菜单语法，之前测试都没发觉这一点

**Reply by tsl0922** [reply]
不想写 key 用下划线代替即可

**Reply by hooke007** [reply]
那下划线这个绑定怎么办，mpvnet的那个省key方式我觉得并不好

**Reply by dyphire** [reply]
太久没用 mpv.net 了，都忘记它的菜单语法了。是的，mpv.net 以下划线来实现无 key 的菜单，uosc 以`#`注释来实现无 key 的菜单
我实际上认为 uosc 的无 key 的菜单语法才是正确的，因为采用注释的写法才不影响脚本不存在时 mpv 的正常使用。而采用 mpv.net 的下划线写法注定这个插件无法合并至 mpv 上游了，不会被接受
 
顺便确实需要实现`input-conf`功能

**Reply by tsl0922** [reply]
已支持完整 uosc 菜单语法。

**Reply by tsl0922** [reply]
input-conf 配置也已经支持。

**Reply by dyphire** [reply]
感觉离初步完善就差轨道菜单和高分屏优化了，我看你之前提到还要实现打开文件功能

**Reply by dyphire** [reply]
目前 4k 屏下的菜单大小看久了也习惯了（
对我来说最大的痛点还不是不支持轨道菜单而是重复性操作体验不够好，执行一些无 key 的重复性操作不如 https://github.com/tsl0922/mpv-plugin/issues/3 中 uosc 演示的那样舒适，毕竟键位是有限的，不可能为所有重复性操作分配键位
不过目前 win 上好像没见过类似 uosc 那样的菜单行为，也不知道能不能实现。linux 上是支持这种菜单行为的

现在的方案是将 uosc 的菜单绑定到 MENU 键位，右键留给菜单插件，使用 uosc 菜单来执行重复性操作

**Reply by tsl0922** [reply]
已初步实现轨道菜单支持，除此之外还支持章节和音频设备选择菜单。

https://github.com/tsl0922/mpv-plugin/pull/13

**Reply by zhongfly** [reply]
> 轨道菜单

![image](https://github.com/hooke007/MPV_lazy/assets/11155705/a68f1fdc-a1d2-469f-8025-495e0de69874)
能不能在没有相应轨道时，子菜单显示一个“无”之类提示的项？我个人看法，毕竟菜单用 > 示意可以展开，结果却没有发生任何反馈，也有点奇怪。

**Reply by dyphire** [reply]
有几个地方需要完善：
- 缺少 editions 菜单
- 轨道标题应该匹配清除当前文件名，参考之前在 mpv.net 中已经实现的：https://github.com/mpvnet-player/mpv.net/pull/432 ，还有其他优化：https://github.com/mpvnet-player/mpv.net/pull/433 ，https://github.com/mpvnet-player/mpv.net/pull/434 。基本就是应该把 mpv.net 已有的轨道菜单上的优化处理复制过来
- 像 mpv.net 那样增加一个 Tracks 菜单能将视频音频字幕和 editions 这些菜单汇总在一起会更好

**Reply by tsl0922** [reply]
这样不太好搞多语言。。所以没显示东西

**Reply by dyphire** [reply]
那当轨道章节类菜单不存在实际内容时可以灰色化处理

**Reply by zhongfly** [reply]
我觉得可以像uosc，左侧显示轨道标题，右对齐显示一些轨道信息
显示的轨道信息能不能自定义，语法类似与`#@tracks/video@lang,codec`这样，按照里track的属性名就行，然后不存在的属性就跳过不显示

**Reply by dyphire** [reply]
感觉支持自定义语法有点复杂了，不太好。右对齐显示轨道信息的想法很好

**Reply by tsl0922** [reply]
还是保持简单点吧，这些用 C 实现太麻烦了。

像是除了标题外的文件格式，文件名之类的信息也不太想搞。太复杂了，这类更详细的信息你需要的是 [MediaInfo](https://mediaarea.net/en/MediaInfo)。

**Reply by dyphire** [reply]
我觉得对外部字幕和音频剥离相同的播放文件名还有右对齐轨道信息的处理还是有必要的。顺便 mpv 已有的轨道信息目前也缺失了很多，这不需要 [MediaInfo](https://mediaarea.net/en/MediaInfo) 就能获取
轨道菜单缺少了很多轨道信息，章节菜单缺少了时间戳信息，还有尚未实现的 editions 版本菜单

**Reply by Youxikong** [reply]
感觉是不是可以增加显示properties的值的功能。比如设置某个快捷键：
`ctrl+h    cycle ontop    #menu: Ontop status: ${ontop}`
菜单就自动把${ontop}替换为这个property的值。

**Reply by natural-harmonia-gropius** [reply]
> 还是保持简单点吧，这些用 C 实现太麻烦了。

能不能让这个项目变成纯粹的菜单显示，尽可能简化这个 dll。菜单内容监听 user-data/contextmenu，具体内容显示什么用其他脚本来完成，比如读取 conf，显示轨道，国际化，字符串超长省略全部交给其他脚本。菜单本身的开关，亮暗也靠像是 user-data/contextmenu/enabled 和 user-data/contextmenu/color-scheme 这样来完成。

**Reply by zhongfly** [reply]
> 具体内容显示什么用其他脚本来完成

这样也不错，专注于只有C plugin才能做到的事，然后其余交给别的语言

**Reply by dyphire** [reply]
我看到目前的章节菜单代码里有 time 的 item，为什么菜单渲染的时候没有显示时间戳？
顺便时间戳我觉得应该使用 [] 包裹

**Reply by tsl0922** [reply]
time 大概是忘记了吧。。可能需要格式化比较麻烦测试的时候没弄上去，加上了。

**Reply by tsl0922** [reply]
> 有点离题：有没有可能之后把ImPlay里的那个调试/信息工具独立出来？

@zhongfly 已单独把 ImPlay 里的调试工具弄出来做成一个 cplugin：https://github.com/tsl0922/mpv-debug-plugin

使用方法和 menu 插件类似，把 dll 文件放到 scripts 目录，然后 `script-message-to debug show` 可以显示 debug 窗口。

**Reply by tsl0922** [reply]
可以本地编译测试，要分发 dll 出去我还得整个静态编译出来的版本才行。

**Reply by dyphire** [reply]
menu plugin 发布 1.0 后，我个人觉得轨道菜单的显示信息还不够完善，自己修改补充了下：https://github.com/dyphire/mpv-menu-plugin/commit/e2317c6e6c391eb6d5e43011aa3e316abff6b06e
显示效果：
![image](https://github.com/hooke007/MPV_lazy/assets/61936050/bfeb2c9f-89da-4acf-9e8b-66d4dd6f2c1d) 
![image](https://github.com/hooke007/MPV_lazy/assets/61936050/820da45d-2069-4024-816f-ce28754eec07)
当然 1.0 已经没必要折腾了，计划中 2.0 的模块化理念明显更优秀

**Comment by hooke007** [comment]
此帖终结
